File: src/tests/jaslib_tests/fixtures/collection_factory.py
from __future__ import annotations

import unittest.mock
from contextlib import contextmanager
from typing import TYPE_CHECKING

from jaslib.note.kanjinote import KanjiNote
from jaslib.note.sentences.sentencenote import SentenceNote
from jaslib_tests.fixtures.base_data.sample_data import kanji_spec, sentence_spec, vocab_lists

if TYPE_CHECKING:
    from collections.abc import Iterator

    from jaslib.note.collection.jp_collection import JPCollection

@contextmanager
def inject_empty_collection() -> Iterator[JPCollection]:
    from jaslib.note.collection.jp_collection import JPCollection
    jp_collection: JPCollection
    def get_jp_collection() -> JPCollection: return jp_collection


    jp_collection = JPCollection()
    with unittest.mock.patch("jaslib.app.col", new=get_jp_collection):
        yield jp_collection
        jp_collection.destruct_sync()


@contextmanager
def inject_collection_with_select_data(kanji: bool = False, special_vocab: bool = False, sentences: bool = False) -> Iterator[JPCollection]:
    with inject_empty_collection() as collection:
        if kanji:
            for _kanji in kanji_spec.test_kanji_list:
                KanjiNote.create(_kanji.question, _kanji.answer, _kanji.on_readings, _kanji.kun_reading)

        if special_vocab:
            for vocab in vocab_lists.test_special_vocab:
                vocab.create_vocab_note()

        if sentences:
            for sentence in sentence_spec.test_sentence_list:
                SentenceNote.create_test_note(sentence.question, sentence.answer)

        yield collection

@contextmanager
def inject_collection_with_all_sample_data() -> Iterator[JPCollection]:
    with inject_collection_with_select_data(kanji=True, special_vocab=True, sentences=True) as collection:
        yield collection

File: src/src_dotnet/JAStudio.Core.Specifications/Fixtures/CollectionFactory.cs
using System;
using JAStudio.Core.Note;
using JAStudio.Core.Note.Sentences;
using JAStudio.Core.Specifications.Fixtures.BaseData.SampleData;

namespace JAStudio.Core.Specifications.Fixtures;

public enum DataNeeded
{
   None = 0,
   Kanji = 2,
   Vocabulary = 4,
   Sentences = 8,
   All = Kanji | Vocabulary | Sentences
}

public static class CollectionFactory
{
   public static AppScope InjectCollectionWithSelectData(DataNeeded data)
   {
      var app = AppBootstrapper.BootstrapForTests();
      if(data == DataNeeded.None)
         return new AppScope(app);

      var noteServices = app.Services.NoteServices;

      if(data.HasFlag(DataNeeded.Kanji))
      {
         foreach(var kanjiSpec in KanjiSpec.TestKanjiList)
         {
            KanjiNote.Create(noteServices, kanjiSpec.Question, kanjiSpec.Answer, kanjiSpec.OnReadings, kanjiSpec.KunReading);
         }
      }

      if(data.HasFlag(DataNeeded.Vocabulary))
      {
         foreach(var vocab in VocabLists.TestSpecialVocab)
         {
            app.Services.VocabNoteFactory.Create(vocab.DisambiguationName, vocab.Answer, vocab.Readings, vocab.InitializeNote);
         }
      }

      if(data.HasFlag(DataNeeded.Sentences))
      {
         foreach(var sentence in SentenceSpec.TestSentenceList)
         {
            SentenceNote.CreateTestNote(noteServices, sentence.Question, sentence.Answer);
         }
      }

      return new AppScope(app);
   }

   public class AppScope(CoreApp coreApp) : IDisposable
   {
      public CoreApp CoreApp { get; } = coreApp;

      public void Dispose()
      {
         CoreApp.Dispose();
      }
   }
}
